subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotterTest.groovy

```diff
@@ -243,6 +253,178 @@ class DefaultFileSystemSnapshotterTest extends Specification {
         })
     }
 
+    def "snapshots a singletonFileTree as RegularFileSnapshot"() {
+        given:
+        def tempDir = tmpDir.createDir('tmpDir')
+        def file = tempDir.file('testFile')
+        file.text = "content"
+
+        when:
+        def tree = new FileTreeAdapter(new DefaultSingletonFileTree(file), TestFiles.patternSetFactory)
+        def snapshots = snapshotter.snapshot(tree)
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        def includedTree = tree.matching(new Action<PatternFilterable>() {
+            @Override
+            void execute(PatternFilterable patternFilterable) {
+                patternFilterable.include("**")
+            }
+        })
+        snapshots = snapshotter.snapshot(includedTree)
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        def excludedTree = tree.matching(new Action<PatternFilterable>() {
+            @Override
+            void execute(PatternFilterable patternFilterable) {
+                patternFilterable.exclude("**")
+            }
+        })
+        snapshots = snapshotter.snapshot(excludedTree)
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        def singleFileTree = TestFiles.fileOperations(tempDir).fileTree(file)
+        snapshots = snapshotter.snapshot(singleFileTree)
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        snapshots = snapshotter.snapshot(singleFileTree.matching { true })
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        snapshots = snapshotter.snapshot(singleFileTree.matching { false })
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        def fromConfigurableFiles = TestFiles.fileOperations(tempDir).configurableFiles(file).asFileTree
+        snapshots = snapshotter.snapshot(fromConfigurableFiles)
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        snapshots = snapshotter.snapshot(fromConfigurableFiles.matching { true })
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+
+
+        when:
+        snapshots = snapshotter.snapshot(fromConfigurableFiles.matching { false })
+
+        then:
+        assertSingleFileSnapshot(snapshots)
+    }
+
+    def "snapshots a archive trees as RegularFileSnapshot"() {
```

Comment:
I guess we should have another test that `tarTree` doesn't blow up when there is no backing file.